text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+* Step 5: Validation
+
+	![Validate](../images/fpc/high-level/FPC-Validate.png)
+
+	As the Peers in the Channel receive the block of transactions, they perform the standard Fabric validation process.
+	In addition, a custom validation plugin in the Peer is responsible to verify FPC transactions. In particular, the FPC Validator queries its local FPC Registry to retrieve the enclave signature verification key and check that the signature was produced in an actual and correct Chaincode Enclave. This query retrieves the stored Attestation report associated with the Public Key of the Chaincode Enclave that produced and signed the transaction. The Attestation report is checked to verify the details of the Chaincode Enclave, affirming the validity of the Enclave.
+
+* Step 6: Trusted Ledger Revalidation
+
+	![Revalidate](../images/fpc/high-level/FPC-Revalidate.png)
+
+	In addition to the standard validation and commit process, the Peer also forwards all committed blocks to the Ledger Enclave in order to establish a full and current trusted view of the ledger. 
+	The same validation steps described above are repeated inside the Enclave, and then the transaction is committed to the trusted version of the Ledger inside this Enclave. This constitutes an update to the World State Integrity Metadata: for each Key Value Pair of World State, a second Key Value Pair is stored in the Trusted Ledger containing Integrity Metadata (a cryptographic hash value) along with channel-specific details needed to verify that the transaction was produced by a valid authorized participant in the channel.
+
+* Step 7: Commitment
+
+	![Commit](../images/fpc/high-level/FPC-Commit.png)
+
+	If these validation processes succeed, the FPC transaction is marked valid.
+	The Peer continues with committing the transactions to the local ledger and applies all valid transactions (i.e. the write set) to the World State. This completes the transaction.
+
+
+## Subsequent Query and Validation of World State Data by FPC Chaincodes
+
+Subsequent to the transaction described above: When any FPC Chaincode accesses the World State, the FPC Shim is responsible not only for passing the query to the Peer but also querying the Ledger Enclave for the Integrity Metadata for the Key in question. It performs a cryptographic hash of the Key Value Pair of World State returned by the Peer and checks it against the Integrity Metadata from the Ledger Enclave for that key. If the hashes do not match, it detects an integrity violation.
+
+## Encrypted Elements
+
+![Encryption](../images/fpc/high-level/FPC-Encryption.png)
+
+Encrypted elements of the FPC architecture (over and above those in Fabric, such as TLS tunnels from Client to Peer) include:
```

Comment:
The below enumerations are extremely high level and obscure and to reason about security, i think it should be explained:

- What keys (symmetric, asymmetric) encrypt the data
- Who owns the keys
- How and where are the keys generated
- How are the keys being distributed

Reply:
The UML diagrams referenced further down have all the corresponding details.

Reply:
I can't understand [this](https://raw.githubusercontent.com/mbrandenburger/fabric-rfcs/fpc-rfc/images/fpc/full-detail/fpc-key-dist.png). The zoom level in which the text is big enough to be readable, prevents you from seeing anything but a small section of the diagram so you lose context. If you zoom out, the text becomes unreadable and then you don't understand anything.

Reply:
(see below)